pkg/connection/manager.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2022 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package connection
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/connection/store"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+// Error strings.
+const (
+	errConnectStore    = "cannot connect to secret store"
+	errWriteStore      = "cannot write to secret store"
+	errDeleteFromStore = "cannot delete from secret store"
+	errGetStoreConfig  = "cannot get store config"
+)
+
+// StoreBuilderFn is a function that builds and returns a Store with a given
+// store config.
+type StoreBuilderFn func(ctx context.Context, local client.Client, cfg v1.SecretStoreConfig) (Store, error)
+
+// A DetailsManagerOption configures a DetailsManager.
+type DetailsManagerOption func(*DetailsManager)
+
+// WithLogger specifies how the DetailsManager should log messages.
+func WithLogger(l logging.Logger) DetailsManagerOption {
+	return func(m *DetailsManager) {
+		m.log = l
+	}
+}
+
+// WithStoreBuilder configures the StoreBuilder to use.
+func WithStoreBuilder(sb StoreBuilderFn) DetailsManagerOption {
+	return func(m *DetailsManager) {
+		m.storeBuilder = sb
+	}
+}
+
+// DetailsManager is a connection details manager that satisfies the required
+// interfaces to work with connection details by managing interaction with
+// different store implementations.
+type DetailsManager struct {
+	client       client.Client
+	newConfig    func() StoreConfig
+	storeBuilder StoreBuilderFn
+
+	log logging.Logger
+}
+
+// NewDetailsManager returns a new connection DetailsManager.
+func NewDetailsManager(c client.Client, of schema.GroupVersionKind, o ...DetailsManagerOption) *DetailsManager {
+	nc := func() StoreConfig {
+		return resource.MustCreateObject(of, c.Scheme()).(StoreConfig)
+	}
```

Comment:
```suggestion
	}
	
	// Panic early if we've been asked to reconcile a resource kind that has not
	// been registered with our controller manager's scheme.
	_ = nc()
```

This is maybe overkill, but we tend to include this when we use `resource.MustCreateObject` just to make sure we panic at startup rather than during the first reconcile.